File: Objective-C/FlickrSearch/FlickrPhotoCell.m
//
//  FlickrPhotoCell.m
//  t5
//
//  Created by Alfred Yu on 2020/10/19.
//

#import "FlickrPhotoCell.h"
#import "FlickrPhoto.h"

@interface FlickrPhotoCell () <FlickrPhoto_Delegate>

@property (strong, nonatomic) FlickrPhoto *photoInfo;

@end

@implementation FlickrPhotoCell

- (void)awakeFromNib
{
    [super awakeFromNib];
    
    [self setLoadingComplete:NO];
    [self imageViewInit];
}

- (void) imageViewInit
{
    [self.imageView setClipsToBounds:YES];
    [self.imageView.layer setCornerRadius:9];
    [self.imageView setContentMode:UIViewContentModeScaleAspectFill];
}

- (void) setLoadingComplete: (BOOL)status
{
    [self.imageView setHidden:(status ? NO : YES)];
    [self.activityView setHidden:(status ? YES : NO)];
    (status ? [self.activityView stopAnimating] : [self.activityView startAnimating]);
}

- (void) loadImageWithInfo: (FlickrPhoto *)info
{
    self.photoInfo = info;
    
    if ( info.image == nil )
    {
        [self setLoadingComplete:NO];
        
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^
        {
            [self.photoInfo loadImage];
            self.photoInfo.delegate = self;
        });
    }
    else
    {
        [self setPhoto];
    }
}

- (void) setPhoto
{
    [self.imageView setImage:self.photoInfo.image];
    [self setLoadingComplete:YES];
}

#pragma mark - FlickrPhoto_Delegate

- (void) photoLoadComplete
{
    dispatch_async(dispatch_get_main_queue(), ^
    {
        [self setPhoto];
    });
}

@end

File: Swift/FlickrSearch/FlickrPhotoCell.swift
//
//  FlickrPhotoCell.swift
//  FlickrSearch
//
//  Created by Alfred Yu on 2020/10/21.
//

import UIKit

class FlickrPhotoCell: UICollectionViewCell, FlickrPhoto_Delegate
{
    @IBOutlet weak var label_id: UILabel!
    @IBOutlet weak var label_owner: UILabel!
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var activityView: UIActivityIndicatorView!
    var photoInfo: FlickrPhoto?
    
    override func awakeFromNib()
    {
        super.awakeFromNib()
        
        setLoadingComplete(status: false)
        imageViewInit()
    }

    public func loadImage(photoinfo: FlickrPhoto)
    {
        self.photoInfo = photoinfo;
        
        if photoinfo.image == nil
        {
            setLoadingComplete(status: false)
            
            DispatchQueue.global(qos: .userInitiated).async
            {
                self.photoInfo?.loadImage()
                self.photoInfo?.delegate = self
            }
        }
        else
        {
            setPhoto()
        }
    }
    
    private func imageViewInit()
    {
        imageView.clipsToBounds = true
        imageView.layer.cornerRadius = 9
        imageView.contentMode = .scaleAspectFill
    }
    
    private func setLoadingComplete(status: Bool)
    {
        if status
        {
            imageView.isHidden = false
            activityView.isHidden = true
            activityView.stopAnimating()
        }
        else
        {
            imageView.isHidden = true
            activityView.isHidden = false
            activityView.startAnimating()
        }
    }
    
    private func setPhoto()
    {
        imageView.image = photoInfo?.image
        setLoadingComplete(status: true)
    }
    
    public func photoLoadComplete()
    {
        DispatchQueue.main.async
        {
            self.setPhoto()
        }
    }
}
